### PR TITLE
facepile ( •_•) ( •_•)>⌐■-■ (⌐■_■)

### PR DESCRIPTION
**Pull request by eoji**
## what
Implementing ✨ facepile ✨ designs showing up to 3 friends' names and avatars.
`SocialUtils` got hairy. Open to any and all suggestions.

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java [hunk 1]
```diff
@@ -131,12 +133,25 @@ public static String friendBackingActivityTitle(final @NonNull Context context,
...
+    final String secondFriendName;
+    if (friends.size() >= 2) {
+      secondFriendName = friends.size() == 2 ? friends.get(1).name() : friends.get(2).name();
+    } else {
+      secondFriendName = "";
+    }
+    final String remainingCount = NumberUtils.format(Math.max(0, friends.size() - 3));
 
-    return ksString.format("discovery_baseball_card_social_friends_are_backers", friends.size(),
+    return ksString.format("discovery_baseball_card_social_friends_are_backers", friends.size() == 3 ? friends.size() - 1 : friends.size(),
```
Comment (swoopej) [inline app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java]: yeah this is all a bit gnarly but it makes sense! If I have any refactoring ideas I'll chime in.

File: app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java [hunk 2]
```diff
@@ -131,12 +133,25 @@ public static String friendBackingActivityTitle(final @NonNull Context context,
   /**
    * Returns a namepile for a list of friends.
    */
-  public static @NonNull String projectCardFriendNamepile(final @NonNull List<User> friends, final @NonNull KSString ksString) {
-    final String friendName = friends.size() >= 1 ? friends.get(0).name() : "";
-    final String secondFriendName = friends.size() >= 2 ? friends.get(1).name() : "";
-    final String remainingCount = NumberUtils.format(Math.max(0, friends.size() - 2));
+  public static @NonNull String projectCardFriendNamepile(final Context context, final @NonNull List<User> friends, final @NonNull KSString ksString) {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java]: `@NonNull` Context

File: app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java [hunk 3]
```diff
@@ -131,12 +133,25 @@ public static String friendBackingActivityTitle(final @NonNull Context context,
...
+    final String friendName;
+    if (friends.size() < 3) {
+      friendName = friends.size() >= 1 ? friends.get(0).name() : "";
+    } else {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/libs/utils/SocialUtils.java]: this logic makes sense, but is a little hard to read through. maybe it'd be helpful to leave a lil comment here with an example of what `friendName` would be set here

File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectCardViewHolder.java [hunk 4]
```diff
@@ -118,10 +120,30 @@ public ProjectCardViewHolder(final @NonNull View view, final @NonNull Delegate d
       .compose(observeForUI())
       .subscribe(ViewUtils.setGone(this.featuredViewGroup));
 
-    this.viewModel.outputs.friendAvatarUrl()
+    this.viewModel.outputs.friendAvatar2IsHidden()
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/ui/viewholders/ProjectCardViewHolder.java]: can rename these outputs from `IsHidden` to `IsGone` to more properly reflect what's happening in the view!

File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectCardViewHolder.java [hunk 5]
```diff
@@ -282,10 +304,10 @@ private void setDeadlineCountdownText(final @NonNull Project project) {
     this.deadlineCountdownUnitTextView.setText(ProjectUtils.deadlineCountdownDetail(project, context(), this.ksString));
   }
 
-  private void setFriendAvatarUrl(final @NonNull String avatarUrl) {
+  private void setFriendAvatarUrl(final @NonNull String avatarUrl, final @NonNull ImageView imageView) {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/ui/viewholders/ProjectCardViewHolder.java]: nice nice

File: app/src/main/res/layout/friend_row_backing_group.xml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  </FrameLayout>
+
+  <TextView
+    android:id="@+id/friend_backing_message"
+    style="@style/Caption1Secondary"
+    android:layout_gravity="center_vertical"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    tools:text="Dancing Hot Dog, Frank, and 1 more are backers." />
```
Comment (luoser) [inline app/src/main/res/layout/friend_row_backing_group.xml]: 🌭

**Comment by eoji** [comment]
[More refresh](https://trello.com/c/Bb0vAdOx/293-more-refresh)

File: app/src/test/java/com/kickstarter/viewmodels/ProjectCardHolderViewModelTest.java [hunk 7]
```diff
@@ -146,15 +155,51 @@ public void testFeaturedViewGroupIsGone_isFeatured() {
...
+      .friends(Arrays.asList(UserFactory.user(), UserFactory.user(), UserFactory.user()))
+      .build();
+    setUpEnvironment(environment());
+
+    this.vm.inputs.configureWith(project);
+    this.friendAvatarUrl1.assertValues(project.friends().get(0).avatar().small());
+    this.friendAvatarUrl2.assertValues(project.friends().get(1).avatar().small());
+    this.friendAvatarUrl3.assertValues(project.friends().get(2).avatar().small());
+    this.friendAvatar2IsHidden.assertValue(false);
+    this.friendAvatar3IsHidden.assertValue(false);
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/test/java/com/kickstarter/viewmodels/ProjectCardHolderViewModelTest.java]: I think `testEmitsFriendBackingViewIsHidden` test handles that case! 🕵️‍♀️

[the remark this answers is not in the document]
Comment (swoopej) [reply app/src/test/java/com/kickstarter/viewmodels/ProjectCardHolderViewModelTest.java]: oh my bad ok!

**Comment by eoji** [comment]
@swoopej @luoser Added some comments to `projectCardFriendNamepile`. Please let me know if they make sense!

**Comment by luoser** [comment]
lgtm! :shipit:

**Comment by swoopej** [comment]
👍 comments are helpful!

**Pull request merged**